lightning/src/util/persist.rs

```diff
@@ -243,10 +250,889 @@ where
+		for update_id in first_update_id..=last_update_id {
+			let sub_ns = monitor_name.as_str();
+			let update_name = UpdateName::from(update_id);
+			debug_assert_eq!(
+				self.read_monitor_update(monitor_name, &update_name)?
+					.update_id,
+				update_name.0
+			);
+			self.kv_store.remove(
+				CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE,
+				sub_ns,
+				update_name.as_str(),
+				true,
+			)?;
+		}
+		Ok(())
+	}
+
+	/// Cleans up stale updates for all monitors, starting at a given search depth.
+	///
+	/// Usually, the value given for `maximum_pending_updates` in [`MonitorUpdatingPersister::new`]
+	/// is the best `search_depth` to use for routine cleaning, such as at startup, as stale updates
+	/// to that depth can be expected if there was a panic and crash.
+	///
+	/// "Deep cleaning" may be a useful, periodic maintenance task. Provide
+	/// [`CLOSED_CHANNEL_UPDATE_ID`] as the `search_depth` to
+	/// clean all updates below each monitor.
+	pub fn cleanup_stale_updates(&self, search_depth: u64) -> io::Result<()>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+	{
+		let monitor_list = self.kv_store.list(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+		)?;
+		for maybe_monitor_name in monitor_list.into_iter().map(MonitorName::new) {
+			let monitor_name = maybe_monitor_name?;
+			let (_, monitor) = self.read_monitor(&monitor_name)?;
+			let end = monitor.get_latest_update_id();
+			let start = end.saturating_sub(search_depth);
+			self.lazy_delete_updates(&monitor_name, start, end)?;
+		}
+		Ok(())
+	}
+}
+
+impl<
+		ChannelSigner: WriteableEcdsaChannelSigner,
+		K: Deref,
+		L: Deref,
+		ES: Deref + Clone,
+		SP: Deref + Clone,
+	> Persist<ChannelSigner> for MonitorUpdatingPersister<K, L, ES, SP>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+	ES::Target: EntropySource + Sized,
+	SP::Target: SignerProvider + Sized,
+{
+	/// Persists a new channel. This means writing the entire monitor to the
+	/// provided [`KVStore`].
+	fn persist_new_channel(
+		&self,
+		funding_txo: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+		_monitor_update_call_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		// Determine the proper key for this monitor
+		let monitor_name = match MonitorName::try_from(funding_txo) {
+			Ok(n) => n,
+			Err(_) => return chain::ChannelMonitorUpdateStatus::InProgress,
+		};
+		let key = monitor_name.as_str();
+		// Check that this key isn't already storing a monitor with a higher update_id (collision)
+		let maybe_old_monitor = self.read_monitor(&monitor_name);
+		match maybe_old_monitor {
+			Ok((_, ref old_monitor)) => {
+				if old_monitor.get_latest_update_id() > monitor.get_latest_update_id() {
+					return chain::ChannelMonitorUpdateStatus::PermanentFailure
+				}
+			},
+			Err(ref e) if e.kind() == io::ErrorKind::NotFound => {},
+			_ => return chain::ChannelMonitorUpdateStatus::InProgress
```

Comment:
How would a user recover from this? Doesn't `InProgress` indicate the operation is happening in the background? But by returning early we've lost the monitor.

Reply:
Initially, I changed these away from `PermanentFailure` at the advice of @TheBlueMatt. I'm not sure they're right now. I haven't come back to these return values since #2562 entered some uncertainty in the feedback phase, and I'm really waiting for both the implementation and documentation guidance there to settle before rebasing and updating to however that shakes out. So really, these are placeholders.

Reply:
Does ever returning `InProgress` rely on having an async `KVStore`? And if so, how do we envision it working with this?

Reply:
More of a question for @TheBlueMatt I think? I don't know. Previously, these implementations only used `Completed` and `PermanentFailure` variants, which more or less worked like `Result` without carried values. I'm unsure if that should translate to the newer API with a find-and-replace, or there are actually places `InProgress` is appropriate here.

Reply:
all of these are UnrecoverableError now